### ISSUE TITLE
Add support for SO_REUSEPORT

### ISSUE THREAD

**Issue by haata** [capnproto/capnproto#1113]
For systems that support it, use SO_REUSEPORT instead of SO_REUSEADDR for 
`Own<ConnectionReceiver> listen()` and `Own<DatagramPort> bindDatagramPort()`.

- Linux >= 3.9
- macOS
- freebsd

https://stackoverflow.com/a/14388707
https://lwn.net/Articles/542629/

**Comment by kentonv** [comment]
FWIW, `SO_REUSEADDR` and `SO_REUSEPORT` do entirely different things; one is not a replacement for the other. We use `SO_REUSEADDR` in the current code because otherwise the port often remains unavailable for 5 minutes after the socket is closed, which is extremely bad for just about every imaginable server use case -- enabling `SO_REUSEADDR` is therefore a no-brainer.

I'm hesitant to add dedicated APIs for exotic socket options. I guess the reason the `ConnectionReceiver::setsockopt()` escape hatch doesn't work here is because the option needs to be set before binding. Maybe we could create some general way to be able to call `setsockopt()` pre-bind?

**Comment by haata** [comment]
Yeah, just having a way to set pre-bind socket options would make it a lot easier.
